In [3]:
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
import pandas as pd
import time

**Iniciando o Driver**

In [2]:
driver = webdriver.Chrome()
driver.get("https://www.transfermarkt.com.br/campeonato-brasileiro-serie-a/gesamtspielplan/wettbewerb/BRA1/saison_id/2024")

**Capturando dados básicos de jogos: equipes e placar**

In [3]:
time_casa = pd.Series([])
gols_casa = pd.Series([])
gols_visitante = pd.Series([])
time_visitante = pd.Series([])

In [4]:
dados_jogos = driver.find_elements(By.CLASS_NAME, "hauptlink")

indice = 0
while indice < len(driver.find_elements(By.CLASS_NAME, "hauptlink")):
    #capturando time da casa
    casa = dados_jogos[indice].find_element(By.TAG_NAME, "a")
    time_casa = pd.concat([time_casa, pd.Series([casa.accessible_name])], ignore_index=True)
    indice = indice + 1

    #capturando placar
    placar_jogo = dados_jogos[indice].find_element(By.TAG_NAME, "a")
    gols_casa = pd.concat([gols_casa, pd.Series([placar_jogo.accessible_name[0]])], ignore_index=True)
    gols_visitante = pd.concat([gols_visitante, pd.Series([placar_jogo.accessible_name[2]])], ignore_index=True)
    indice = indice + 1

    #capturando time visitante
    visitante = dados_jogos[indice].find_element(By.TAG_NAME, "a")
    time_visitante = pd.concat([time_visitante, pd.Series([visitante.accessible_name])], ignore_index=True)
    indice = indice + 1

In [5]:
resultados = pd.DataFrame({
    "time_casa": time_casa,
    "gols_casa": gols_casa,
    "gols_visitante": gols_visitante,
    "time_visitante": time_visitante
})

In [6]:
# pd.Series([int(resultados['gols_casa'])+int(resultados['gols_visitante'])])
resultados[-10:]

time_casa gols_casa gols_visitante time_visitante
370     Fluminense         -              -          Bahia
371       Botafogo         -              -      Fortaleza
372    Corinthians         -              -      Juventude
373         Santos         -              -       Cruzeiro
374       Mirassol         -              -       Flamengo
375    Atlético-MG         -              -  Vasco da Gama
376  Internacional         -              -     Bragantino
377        Vitória         -              -      São Paulo
378          Ceará         -              -      Palmeiras
379          Sport         -              -         Grêmio

In [7]:
#convertendo valor de placar para numérico e placar de jogos não realizados para NaN

resultados['gols_casa'] = pd.to_numeric(resultados['gols_casa'], errors='coerce')
resultados['gols_visitante'] = pd.to_numeric(resultados['gols_visitante'], errors='coerce')

In [8]:
#criação de coluna com o total de gols da partida

resultados['gols_total'] = resultados['gols_casa'] + resultados['gols_visitante']

In [9]:
#média de gols do São Paulo dentro e fora de casa

media_gols_sp_casa = resultados[resultados['time_casa'] == 'São Paulo']['gols_casa'].mean()
media_gols_sp_fora = resultados[resultados['time_visitante'] == 'São Paulo']['gols_visitante'].mean()

moda_gols_sp_casa = resultados[resultados['time_casa'] == 'São Paulo']['gols_casa'].mode()[0]
moda_gols_sp_fora = resultados[resultados['time_visitante'] == 'São Paulo']['gols_visitante'].mode()[0]

print("A MÉDIA de gols do Sâo Paulo dentro de casa é {:.2f}".format(media_gols_sp_casa))
print("A MÉDIA de gols do Sâo Paulo fora de casa é {:.2f}".format(media_gols_sp_fora))
print()
print("A MODA de gols do Sâo Paulo dentro de casa é {}".format(moda_gols_sp_casa))
print("A MODA de gols do Sâo Paulo fora de casa é {}".format(moda_gols_sp_fora))

A MÉDIA de gols do Sâo Paulo dentro de casa é 1.33
A MÉDIA de gols do Sâo Paulo fora de casa é 0.94

A MODA de gols do Sâo Paulo dentro de casa é 2.0
A MODA de gols do Sâo Paulo fora de casa é 0.0


In [10]:
#analisando um duelo específico Cruzeiro x Corinthians
#veremos a media e moda dentro de casa do Cruzeiro, assim como a do Corinthians fora de casa

media_gols_cru_casa = resultados[resultados['time_casa'] == 'Bragantino']['gols_casa'].mean()
moda_gols_cru_casa = resultados[resultados['time_casa'] == 'Bragantino']['gols_casa'].mode()[0]

media_gols_cor_fora = resultados[resultados['time_visitante'] == 'Fortaleza']['gols_visitante'].mean()
moda_gols_cor_fora = resultados[resultados['time_visitante'] == 'Fortaleza']['gols_visitante'].mode()[0]

print("A MÉDIA de gols do Bragantino dentro de casa é {:.2f}".format(media_gols_cru_casa))
print("A MODA de gols do Bragantino dentro de casa é {:.2f}".format(moda_gols_cru_casa))
print()
print("A MÉDIA de gols do Fortaleza fora de casa é {:.2f}".format(media_gols_cor_fora))
print("A MODA de gols do Fortaleza fora de casa é {:.2f}".format(moda_gols_cor_fora))

A MÉDIA de gols do Bragantino dentro de casa é 1.28
A MODA de gols do Bragantino dentro de casa é 1.00

A MÉDIA de gols do Fortaleza fora de casa é 1.00
A MODA de gols do Fortaleza fora de casa é 1.00


In [12]:
#analisando um duelo específico Cruzeiro x Corinthians
#veremos a media e moda dentro de casa do Cruzeiro, assim como a do Corinthians fora de casa

media_gols_cru_casa = resultados[resultados['time_casa'] == 'Fluminense']['gols_casa'].mean()
moda_gols_cru_casa = resultados[resultados['time_casa'] == 'Fluminense']['gols_casa'].mode()[0]

media_gols_cor_fora = resultados[resultados['time_visitante'] == 'São Paulo']['gols_visitante'].mean()
moda_gols_cor_fora = resultados[resultados['time_visitante'] == 'São Paulo']['gols_visitante'].mode()[0]

print("A MÉDIA de gols do Fluminense dentro de casa é {:.2f}".format(media_gols_cru_casa))
print("A MODA de gols do Fluminense dentro de casa é {:.2f}".format(moda_gols_cru_casa))
print()
print("A MÉDIA de gols do São Paulo fora de casa é {:.2f}".format(media_gols_cor_fora))
print("A MODA de gols do São Paulo fora de casa é {:.2f}".format(moda_gols_cor_fora))

A MÉDIA de gols do Fluminense dentro de casa é 1.35
A MODA de gols do Fluminense dentro de casa é 1.00

A MÉDIA de gols do São Paulo fora de casa é 0.94
A MODA de gols do São Paulo fora de casa é 0.00


In [13]:
#para encerrar o driver corretamente
driver.quit()

**A Patir de agora será feita a captura detalhada dos dados (escanteio, faltas, finalização, etc).**

In [29]:
from selenium import webdriver
from selenium.webdriver.common.by import By
from selenium.webdriver.support.ui import WebDriverWait
from selenium.webdriver.support import expected_conditions as EC
from selenium.webdriver.chrome.options import Options
from selenium.webdriver.chrome.service import Service
import pandas as pd
import time
import os
import psutil

Captura dos Rótulos dos Dados

In [81]:
chrome_options = Options()

prefs = {"profile.default_content_setting_values.notifications": 2}

chrome_options.add_experimental_option("prefs", prefs)

service = Service(service_args=['--read-timeout=60'])


driver = webdriver.Chrome(options=chrome_options, service=service)
pid = service.process.pid

# driver = webdriver.Chrome(options=chrome_options)
# driver = webdriver.Chrome()
driver.get("https://www.transfermarkt.com.br/campeonato-brasileiro-serie-a/gesamtspielplan/wettbewerb/BRA1/saison_id/2024")

wait = WebDriverWait(driver, 5)
wait.until(EC.frame_to_be_available_and_switch_to_it("sp_message_iframe_954038"))

titulo = driver.title

wait = WebDriverWait(driver, 7)

botao = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "accept-all")))

botao.click()

driver.switch_to.default_content()

time.sleep(3)

In [82]:
pid

8884

In [83]:
parent = psutil.Process(pid)

for child in parent.children(recursive=True):
    try:
        child.kill()
    except psutil.NoSuchProcess:
        pass

parent.kill()

In [ ]:
for child in parente

In [79]:
psutil.Process(pid).terminate(kill_tree=True)

TypeError: Process.terminate() got an unexpected keyword argument 'kill_tree'

In [53]:
process.is_running()

True

In [56]:
driver.quit()

In [54]:

process.kill()

In [43]:
os.kill(pid, 15)

PermissionError: [WinError 5] Acesso negado

In [63]:
#para entrarmos na pagina do jogo
wait = WebDriverWait(driver, 7)
# link_jogo = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "ergebnis-link")))
wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "ergebnis-link")))
# time.sleep(6)
links_jogos = driver.find_elements(By.CLASS_NAME, "ergebnis-link")
links_jogos[0].click()

In [64]:
#para remover o anuncio da tela
try:
    driver.switch_to.frame("google_ads_iframe_/58778164/transfermarkt.com.br/transfermarkt.com.br_d_interstitial_0")
    fechar_anuncio = driver.find_elements(By.CLASS_NAME, "skip")
    fechar_anuncio[0].click()
    driver.switch_to.default_content()
except:
    pass

In [65]:
#para acessarmos a pagina de esttísticas do jogo
pag_est = driver.current_url.replace('index', 'statistik')
driver.get(pag_est)

In [66]:
#para capturarmos as informações dos rótulos/nomes dos dados/estatísticas
rotulos = driver.find_elements(By.CLASS_NAME, "unterueberschrift")
rotulos = rotulos[1:-2] #capturando apenas os dados interessantes
rotulos_desc = []
for indice in range(0, len(rotulos)):
    rotulos_desc.append(rotulos[indice].text)


rotulos = []
for indice in range(0, len(rotulos_desc)):
    rotulos.append(rotulos_desc[indice]+" casa")
    rotulos.append(rotulos_desc[indice]+" visitante")

rotulos.insert(0, "Gols visitante")
rotulos.insert(0, "Gols casa")
rotulos.insert(0, "Time visitante")
rotulos.insert(0, "Time casa")

time.sleep(2)

In [68]:
#funcionalidade para retornar à página inicial
try:
    while(driver.title != titulo): #a variável titulo é o nome da página inicial
        driver.back()
        try:
            WebDriverWait(driver, 5).until( lambda driver: len(driver.title)>0 )
        except:    
            pass
except:
    pass

In [69]:
rotulos

['Time casa',
 'Time visitante',
 'Gols casa',
 'Gols visitante',
 'Tentativas de gol casa',
 'Tentativas de gol visitante',
 'Chutes para fora casa',
 'Chutes para fora visitante',
 'Defesa casa',
 'Defesa visitante',
 'Escanteios casa',
 'Escanteios visitante',
 'Cobranças de falta casa',
 'Cobranças de falta visitante',
 'Faltas casa',
 'Faltas visitante',
 'Impedimentos casa',
 'Impedimentos visitante']

In [48]:
len(dados_jogo)

16

In [49]:
len(rotulos)

18

In [70]:
driver.quit()

-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-+-

Captura dos Dados dos Jogos

In [ ]:
chrome_options = Options()

prefs = {"profile.default_content_setting_values.notifications": 2}

chrome_options.add_experimental_option("prefs", prefs)
chrome_options.page_load_strategy = 'eager'
service = Service(service_args=['--read-timeout=300'])


driver = webdriver.Chrome(options=chrome_options, service=service)
pid = service.process.pid
# driver = webdriver.Chrome(options=chrome_options)
# driver = webdriver.Chrome()
driver.get("https://www.transfermarkt.com.br/campeonato-brasileiro-serie-a/gesamtspielplan/wettbewerb/BRA1/saison_id/2024")

titulo = driver.title

try:
    wait = WebDriverWait(driver, 5)
    wait.until(EC.frame_to_be_available_and_switch_to_it("sp_message_iframe_954038"))

    wait = WebDriverWait(driver, 7)
    botao = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "accept-all")))
    botao.click()

    driver.switch_to.default_content()
except:
    pass
driver.switch_to.default_content()

time.sleep(3)

In [72]:
dataset_jogos = pd.DataFrame(columns=rotulos) #declaração do dataframe para acumular os dados dos jogos

In [73]:
try:
    wait = WebDriverWait(driver, 7)

    botao = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "accept-all")))

    botao.click()

    driver.switch_to.default_content()
except:
    pass

In [75]:
#para entrarmos na pagina do jogo
wait = WebDriverWait(driver, 7)
wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "ergebnis-link")))
links_estatistica = driver.find_elements(By.CLASS_NAME, "ergebnis-link")

In [76]:
tab_atual = driver.current_window_handle

jogo = 0
while jogo<len(links_estatistica):
    try:
        # time.sleep(3)
        link = links_estatistica[jogo].get_attribute('href')
        
        if(jogo%5 == 0):
            print(link, jogo)
            
        driver.switch_to.new_window('tab')
        driver.get(link)

        try:
            driver.switch_to.frame("google_ads_iframe_/58778164/transfermarkt.com.br/transfermarkt.com.br_d_interstitial_0")
            fechar_anuncio = driver.find_elements(By.CLASS_NAME, "skip")
            fechar_anuncio[0].click()
            driver.switch_to.default_content()
        except:
            pass
        driver.switch_to.default_content()

        # time.sleep(2)
        pag_est = driver.current_url.replace('index', 'statistik')

        driver.get(pag_est)

        # time.sleep(2)
        #capturando os times da partida
        times = []
        wait = WebDriverWait(driver, 5)
        wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "sb-vereinslink")))
        dados = driver.find_elements(By.CLASS_NAME, "sb-vereinslink")
        for indice in range(0, len(dados)):
            times.append(dados[indice].text)


        #capturando o placar da partida
        placar = []
        wait = WebDriverWait(driver, 5)
        wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "sb-endstand")))
        dados = driver.find_elements(By.CLASS_NAME, "sb-endstand")
        placar.append(dados[0].text[0])
        placar.append(dados[0].text[2])
        placar = [int(s) for s in placar]


        #catpura de dados estatísticos do jogo
        wait = WebDriverWait(driver, 5)
        wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "sb-statistik-zahl")))
        dados = driver.find_elements(By.CLASS_NAME, "sb-statistik-zahl")
        lista_dados = []
        for indice in range(0, len(dados)):
            lista_dados.append(dados[indice].text)
        lista_dados = [int(s) for s in lista_dados]
        
        #concatenação de todos os dados desse jogo
        dados_jogo = times + placar + lista_dados
        
        #adição dos dados desse jogo ao dataset
        dataset_jogos.loc[len(dataset_jogos)] = pd.Series(dados_jogo).values

        
        driver.close()
        driver.switch_to.window(tab_atual)
        
    except:
        driver.switch_to.window(tab_atual)
        driver.get("https://www.transfermarkt.com.br/campeonato-brasileiro-serie-a/gesamtspielplan/wettbewerb/BRA1/saison_id/2024")
        driver.quit()
        chrome_options = Options()

        prefs = {"profile.default_content_setting_values.notifications": 2}

        chrome_options.add_experimental_option("prefs", prefs)
        chrome_options.page_load_strategy = 'eager'
        service = Service(service_args=['--read-timeout=300'])

        driver = webdriver.Chrome(options=chrome_options, service=service)
        pid = service.process.pid
        # driver = webdriver.Chrome(options=chrome_options)
        # driver = webdriver.Chrome()
        driver.get("https://www.transfermarkt.com.br/campeonato-brasileiro-serie-a/gesamtspielplan/wettbewerb/BRA1/saison_id/2024")

        titulo = driver.title
        tab_atual = driver.current_window_handle

        try:
            wait = WebDriverWait(driver, 5)
            wait.until(EC.frame_to_be_available_and_switch_to_it("sp_message_iframe_954038"))

            wait = WebDriverWait(driver, 7)
            botao = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "accept-all")))
            botao.click()

            driver.switch_to.default_content()
        except:
            pass
        driver.switch_to.default_content()

        time.sleep(2)

        #para entrarmos na pagina do jogo
        wait = WebDriverWait(driver, 7)
        wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "ergebnis-link")))
        links_estatistica = driver.find_elements(By.CLASS_NAME, "ergebnis-link")

        jogo = jogo+1
        continue

    jogo = jogo+1

https://www.transfermarkt.com.br/sao-paulo-fc_sport-recife/index/spielbericht/4582063 0
https://www.transfermarkt.com.br/cr-flamengo_sc-internacional/index/spielbericht/4582060 5


MaxRetryError: HTTPConnectionPool(host='localhost', port=50280): Max retries exceeded with url: /session/908ff3a3a6da0d784e397d85711c3f65/window (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x000001CF6B8F17B0>: Failed to establish a new connection: [WinError 10061] Nenhuma conexão pôde ser feita porque a máquina de destino as recusou ativamente'))

In [62]:
tab_atual = driver.current_window_handle

jogo = 0
while jogo<len(links_estatistica):
    # time.sleep(3)
    link = links_estatistica[jogo].get_attribute('href')
    
    if(jogo%5 == 0):
        print(link, jogo)
        
    driver.switch_to.new_window('tab')
    driver.get(link)

    try:
        driver.switch_to.frame("google_ads_iframe_/58778164/transfermarkt.com.br/transfermarkt.com.br_d_interstitial_0")
        fechar_anuncio = driver.find_elements(By.CLASS_NAME, "skip")
        fechar_anuncio[0].click()
        driver.switch_to.default_content()
    except:
        pass
    driver.switch_to.default_content()

    # time.sleep(2)
    pag_est = driver.current_url.replace('index', 'statistik')

    driver.get(pag_est)

    # time.sleep(2)
    #capturando os times da partida
    times = []
    wait = WebDriverWait(driver, 5)
    wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "sb-vereinslink")))
    dados = driver.find_elements(By.CLASS_NAME, "sb-vereinslink")
    for indice in range(0, len(dados)):
        times.append(dados[indice].text)


    #capturando o placar da partida
    placar = []
    wait = WebDriverWait(driver, 5)
    wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "sb-endstand")))
    dados = driver.find_elements(By.CLASS_NAME, "sb-endstand")
    placar.append(dados[0].text[0])
    placar.append(dados[0].text[2])
    placar = [int(s) for s in placar]


    #catpura de dados estatísticos do jogo
    wait = WebDriverWait(driver, 5)
    wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "sb-statistik-zahl")))
    dados = driver.find_elements(By.CLASS_NAME, "sb-statistik-zahl")
    lista_dados = []
    for indice in range(0, len(dados)):
        lista_dados.append(dados[indice].text)
    lista_dados = [int(s) for s in lista_dados]
    
    #concatenação de todos os dados desse jogo
    dados_jogo = times + placar + lista_dados
    
    #adição dos dados desse jogo ao dataset
    dataset_jogos.loc[len(dataset_jogos)] = pd.Series(dados_jogo).values

    
    driver.close()
    driver.switch_to.window(tab_atual)

    jogo = jogo+1

https://www.transfermarkt.com.br/sao-paulo-fc_sport-recife/index/spielbericht/4582063 0
https://www.transfermarkt.com.br/cr-flamengo_sc-internacional/index/spielbericht/4582060 5
https://www.transfermarkt.com.br/sc-corinthians_cr-vasco-da-gama/index/spielbericht/4582132 10


ReadTimeoutError: HTTPConnectionPool(host='localhost', port=53138): Read timed out. (read timeout=120)

In [63]:
dataset_jogos

Time casa    Time visitante Gols casa Gols visitante  \
0       São Paulo FC      Sport Recife         0              0   
1        Cruzeiro EC       Mirassol FC         2              1   
2        Grêmio FBPA  Atlético Mineiro         2              1   
3       Fortaleza EC     Fluminense FC         2              0   
4       EC Juventude        EC Vitória         2              0   
5        CR Flamengo  SC Internacional         1              1   
6       SE Palmeiras       Botafogo FR         0              0   
7   CR Vasco da Gama         Santos FC         2              1   
8           EC Bahia    SC Corinthians         1              1   
9      RB Bragantino          Ceará SC         2              2   
10    SC Corinthians  CR Vasco da Gama         3              0   

   Tentativas de gol casa Tentativas de gol visitante Chutes para fora casa  \
0                      16                           8                    11   
1                       8                          16                     5   
2                      12                          22                     4   
3                       5                          15                     2   
4                       9                          14                     3   
5                      19                           5                    11   
6                      13                          13                     6   
7                      12                          11                     6   
8                      14                           9                     6   
9                      18                          13                     6   
10                      6                          10                     2   

   Chutes para fora visitante Defesa casa Defesa visitante Escanteios casa  \
0                           4           1                3               6   
1                           5           5                0               2   
2                           5           6                3               1   
3                           9           1                1               4   
4                           3           5                2               3   
5                           1           0                4               4   
6                           5           5                3               2   
7                           5           1                3               2   
8                           3           3                4               3   
9                           6           4                5               9   
10                          6           3                1               1   

   Escanteios visitante Cobranças de falta casa Cobranças de falta visitante  \
0                     2                      20                           16   
1                     6                      20                           18   
2                    11                       5                           17   
3                     5                       5                           15   
4                     5                      16                           16   
5                     2                      16                           10   
6                     2                      22                           17   
7                     3                      11                           11   
8                     5                      15                           15   
9                     2                      13                           13   
10                    4                      15                           18   

   Faltas casa Faltas visitante Impedimentos casa Impedimentos visitante  
0           17               21                 1                      2  
1           18               21                 1                      2  
2           19                5                 1                      4  
3           16                5                 4          

**=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=**

In [50]:
dataset_jogos

Empty DataFrame
Columns: [Time casa, Time visitante, Gols casa, Gols visitante, Tentativas de gol casa, Tentativas de gol visitante, Chutes para fora casa, Chutes para fora visitante, Defesa casa, Defesa visitante, Escanteios casa, Escanteios visitante, Cobranças de falta casa, Cobranças de falta visitante, Faltas casa, Faltas visitante, Impedimentos casa, Impedimentos visitante]
Index: []

In [ ]:
tab_atual = driver.current_window_handle

jogo = 0
while jogo<len(links_estatistica):
    time.sleep(3)
    link = links_estatistica[jogo].get_attribute('href')
    
    if(jogo%5 == 0):
        print(link, jogo)
        
    driver.switch_to.new_window('tab')
    driver.get(link)

    try:
        driver.switch_to.frame("google_ads_iframe_/58778164/transfermarkt.com.br/transfermarkt.com.br_d_interstitial_0")
        fechar_anuncio = driver.find_elements(By.CLASS_NAME, "skip")
        fechar_anuncio[0].click()
        driver.switch_to.default_content()
    except:
        pass
    driver.switch_to.default_content()

    time.sleep(2)
    pag_est = driver.current_url.replace('index', 'statistik')

    driver.get(pag_est)

    time.sleep(2)
    #capturando os times da partida
    times = []
    dados = driver.find_elements(By.CLASS_NAME, "sb-vereinslink")
    for indice in range(0, len(dados)):
        times.append(dados[indice].text)


    #capturando o placar da partida
    placar = []
    dados = driver.find_elements(By.CLASS_NAME, "sb-endstand")
    placar.append(dados[0].text[0])
    placar.append(dados[0].text[2])
    placar = [int(s) for s in placar]


    #catpura de dados estatísticos do jogo
    dados = driver.find_elements(By.CLASS_NAME, "sb-statistik-zahl")
    lista_dados = []
    for indice in range(0, len(dados)):
        lista_dados.append(dados[indice].text)
    lista_dados = [int(s) for s in lista_dados]
    
    #concatenação de todos os dados desse jogo
    dados_jogo = times + placar + lista_dados
    
    #adição dos dados desse jogo ao dataset
    dataset_jogos.loc[len(dataset_jogos)] = pd.Series(dados_jogo).values

    
    driver.close()
    driver.switch_to.window(tab_atual)

    jogo = jogo+1

-----------------------------------------------------------------------------------

In [29]:
dataset_jogos

Time casa    Time visitante Gols casa Gols visitante  \
0       São Paulo FC      Sport Recife         0              0   
1        Cruzeiro EC       Mirassol FC         2              1   
2        Grêmio FBPA  Atlético Mineiro         2              1   
3       Fortaleza EC     Fluminense FC         2              0   
4       EC Juventude        EC Vitória         2              0   
..               ...               ...       ...            ...   
64          EC Bahia       Botafogo FR         1              0   
65  CR Vasco da Gama      SE Palmeiras         0              1   
66       Grêmio FBPA         Santos FC         1              0   
67       Cruzeiro EC       CR Flamengo         2              1   
68     RB Bragantino       Mirassol FC         1              0   

   Tentativas de gol casa Tentativas de gol visitante Chutes para fora casa  \
0                      16                           8                    11   
1                       8                          16                     5   
2                      12                          22                     4   
3                       5                          15                     2   
4                       9                          14                     3   
..                    ...                         ...                   ...   
64                      9                          16                     4   
65                     13                          16                     4   
66                      8                          13                     3   
67                     17                           6                     2   
68                     15                          12                     7   

   Chutes para fora visitante Defesa casa Defesa visitante Escanteios casa  \
0                           4           1                3               6   
1                           5           5                0               2   
2                           5           6                3               1   
3                           9           1                1               4   
4                           3           5                2               3   
..                        ...         ...              ...             ...   
64                          4           4                1               3   
65                         11           1                4               7   
66                          7           1                2               3   
67                          2           3                9               6   
68                          6           3                3               4   

   Escanteios visitante Cobranças de falta casa Cobranças de falta visitante  \
0                     2                      20                           16   
1                     6                      20                           18   
2                    11                       5                           17   
3                     5                       5                           15   
4                     5                      16                           16   
..                  ...                     ...                          ...   
64                    3                       5                           10   
65                   11                       8                           11   
66                    6                      18                            9   
67                    5                       7                           18   
68                    4                      11                            8   

   Faltas casa Faltas visitante Impedimentos casa Impedimentos visitante  
0           17               21                 1                      2  
1           18               21                 1                      2  
2           19                5                 1                      4  
3           16                5                 4          

**=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=**

**=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=+=**

### **TESTES**

In [ ]:
dataset_jogos = pd.DataFrame(columns=rotulos) #declaração do dataframe para acumular os dados dos jogos

try:
    driver.switch_to.frame("sp_message_iframe_954038")

    wait = WebDriverWait(driver, 7)

    botao = wait.until(EC.element_to_be_clickable((By.CLASS_NAME, "accept-all")))

    botao.click()

    driver.switch_to.default_content()
except:
    pass

time.sleep(3)

links_estatistica = driver.find_elements(By.CLASS_NAME, "ergebnis-link")

for jogo in links_estatistica:
    jogo.click()

    try:
        driver.switch_to.frame("google_ads_iframe_/58778164/transfermarkt.com.br/transfermarkt.com.br_d_interstitial_0")
        fechar_anuncio = driver.find_elements(By.CLASS_NAME, "skip")
        fechar_anuncio[0].click()
        driver.switch_to.default_content()
    except:
        pass

    pag_est = driver.current_url.replace('index', 'statistik')

    driver.get(pag_est)

    #capturando os times da partida
    times = []
    dados = driver.find_elements(By.CLASS_NAME, "sb-vereinslink")
    for indice in range(0, len(dados)):
        times.append(dados[indice].text)


    #capturando o placar da partida
    placar = []
    dados = driver.find_elements(By.CLASS_NAME, "sb-endstand")
    placar.append(dados[0].text[0])
    placar.append(dados[0].text[2])
    placar = [int(s) for s in placar]


    #catpura de dados estatísticos do jogo
    dados = driver.find_elements(By.CLASS_NAME, "sb-statistik-zahl")
    lista_dados = []
    for indice in range(0, len(dados)):
        lista_dados.append(dados[indice].text)
    lista_dados = [int(s) for s in lista_dados]
    
    #concatenação de todos os dados desse jogo
    dados_jogo = times + placar + lista_dados
    
    #adição dos dados desse jogo ao dataset
    dataset_jogos.loc[len(dataset_jogos)] = pd.Series(dados_jogo).values

    
    #funcionalidade para retornar à página inicial
    try:
        while(driver.title != titulo): #a variável titulo é o nome da página inicial
            driver.back()
            try:
                WebDriverWait(driver, 5).until( lambda driver: len(driver.title)>0 )
            except:    
                pass
    except:
        pass

In [256]:
dados = driver.find_elements(By.CLASS_NAME, "sb-statistik-zahl")
lista_dados = []
for indice in range(0, len(dados)):
    lista_dados.append(dados[indice].text)

lista_dados = [int(s) for s in lista_dados]
lista_dados

[8, 16, 5, 5, 5, 0, 2, 6, 20, 18, 18, 21, 1, 2]

In [257]:
#capturando os times da partida
times = []
dados = driver.find_elements(By.CLASS_NAME, "sb-vereinslink")
for indice in range(0, len(dados)):
    times.append(dados[indice].text)

times

['Cruzeiro EC', 'Mirassol FC']

In [258]:
#capturando o placar da partida

placar = []
dados = driver.find_elements(By.CLASS_NAME, "sb-endstand")
placar.append(dados[0].text[0])
placar.append(dados[0].text[2])
placar = [int(s) for s in placar]
placar

[2, 1]

In [259]:
dados_totais = times + placar + lista_dados
dados_totais

['Cruzeiro EC',
 'Mirassol FC',
 2,
 1,
 8,
 16,
 5,
 5,
 5,
 0,
 2,
 6,
 20,
 18,
 18,
 21,
 1,
 2]

In [ ]:
dataset_jogos = pd.DataFrame(columns=rotulos)
dataset_jogos.loc[len(dataset_jogos)] = pd.Series(dados_jogo).values
dataset_jogos


Empty DataFrame
Columns: [Time casa, Time visitante, Gols casa, Gols visitante, Tentativas de gol casa, Tentativas de gol visitante, Chutes para fora casa, Chutes para fora visitante, Defesa casa, Defesa visitante, Escanteios casa, Escanteios visitante, Cobranças de falta casa, Cobranças de falta visitante, Faltas casa, Faltas visitante, Impedimentos casa, Impedimentos visitante]
Index: []

In [ ]:
dataset_jogos.loc[len(dataset_jogos)] = pd.Series(dados_jogo).values

In [262]:
teste

Time casa Time visitante Gols casa Gols visitante Tentativas de gol casa  \
0  Cruzeiro EC    Mirassol FC         2              1                      8   

  Tentativas de gol visitante Chutes para fora casa  \
0                          16                     5   

  Chutes para fora visitante Defesa casa Defesa visitante Escanteios casa  \
0                          5           5                0               2   

  Escanteios visitante Cobranças de falta casa Cobranças de falta visitante  \
0                    6                      20                           18   

  Faltas casa Faltas visitante Impedimentos casa Impedimentos visitante  
0          18               21                 1                      2

**Para Encerrar a execução geral, execute a célula abaixo**

In [ ]:
driver.back()
driver.back()

In [179]:
driver.quit()

**TESTES**

In [152]:
esc_bra = [5, 6, 4, 6, 2, 5, 5, 6]

In [154]:
sum(esc_bra)/len(esc_bra)

4.875

In [155]:
esc_for = [6, 5, 2, 1, 6, 4, 2]
sum(esc_for)/len(esc_for)

3.7142857142857144

In [ ]:
for jogo in range(0, len(links_estatistica)):
    # WebDriverWait(driver, 10).until(EC.element_to_be_clickable((By.CLASS_NAME, "ergebnis-link")))
    links = driver.find_elements(By.CLASS_NAME, "ergebnis-link")
    time.sleep(4)
    links[0].click()

    try:
        driver.switch_to.frame("google_ads_iframe_/58778164/transfermarkt.com.br/transfermarkt.com.br_d_interstitial_0")
        fechar_anuncio = driver.find_elements(By.CLASS_NAME, "skip")
        fechar_anuncio[0].click()
        driver.switch_to.default_content()
    except:
        pass
    driver.switch_to.default_content()

    pag_est = driver.current_url.replace('index', 'statistik')

    driver.get(pag_est)

    #capturando os times da partida
    times = []
    dados = driver.find_elements(By.CLASS_NAME, "sb-vereinslink")
    for indice in range(0, len(dados)):
        times.append(dados[indice].text)


    #capturando o placar da partida
    placar = []
    dados = driver.find_elements(By.CLASS_NAME, "sb-endstand")
    placar.append(dados[0].text[0])
    placar.append(dados[0].text[2])
    placar = [int(s) for s in placar]


    #catpura de dados estatísticos do jogo
    dados = driver.find_elements(By.CLASS_NAME, "sb-statistik-zahl")
    lista_dados = []
    for indice in range(0, len(dados)):
        lista_dados.append(dados[indice].text)
    lista_dados = [int(s) for s in lista_dados]
    
    #concatenação de todos os dados desse jogo
    dados_jogo = times + placar + lista_dados
    
    #adição dos dados desse jogo ao dataset
    dataset_jogos.loc[len(dataset_jogos)] = pd.Series(dados_jogo).values

    
    #funcionalidade para retornar à página inicial
    try:
        while(driver.title != titulo): #a variável titulo é o nome da página inicial
            driver.back()
            try:
                WebDriverWait(driver, 5).until( lambda driver: len(driver.title)>0 )
            except:    
                pass
    except:
        pass

    driver.get("https://www.transfermarkt.com.br/campeonato-brasileiro-serie-a/gesamtspielplan/wettbewerb/BRA1/saison_id/2024")

InvalidSessionIdException: Message: invalid session id: session deleted as the browser has closed the connection
from disconnected: not connected to DevTools
  (Session info: chrome=142.0.7444.176); For documentation on this error, please visit: https://www.selenium.dev/documentation/webdriver/troubleshooting/errors#invalidsessionidexception
Stacktrace:
Symbols not available. Dumping unresolved backtrace:
	0x7ff6bd17a235
	0x7ff6bced2630
	0x7ff6bcc616dd
	0x7ff6bcc4d205
	0x7ff6bcc72a8a
	0x7ff6bccea476
	0x7ff6bcd0b3d2
	0x7ff6bccab0ed
	0x7ff6bccabf63
	0x7ff6bd1a5d60
	0x7ff6bd19fe8a
	0x7ff6bd1c1005
	0x7ff6bceed71e
	0x7ff6bcef4e1f
	0x7ff6bcedb7c4
	0x7ff6bcedb97f
	0x7ff6bcec18e8
	0x7fff6789e8d7
	0x7fff692ec53c


In [ ]:
indice = 0
while indice<len(links_estatistica):
    # WebDriverWait(driver, 10).until(EC.element_to_be_clickable((By.CLASS_NAME, "ergebnis-link")))
    WebDriverWait(driver, 5).until(EC.presence_of_all_elements_located((By.CLASS_NAME, "ergebnis-link")))
    links = driver.find_elements(By.CLASS_NAME, "ergebnis-link")
    time.sleep(4)
    links[indice].click()

    try:
        driver.switch_to.frame("google_ads_iframe_/58778164/transfermarkt.com.br/transfermarkt.com.br_d_interstitial_0")
        fechar_anuncio = driver.find_elements(By.CLASS_NAME, "skip")
        fechar_anuncio[0].click()
        driver.switch_to.default_content()
    except:
        pass
    driver.switch_to.default_content()

    pag_est = driver.current_url.replace('index', 'statistik')

    driver.get(pag_est)

    #capturando os times da partida
    times = []
    dados = driver.find_elements(By.CLASS_NAME, "sb-vereinslink")
    for indice in range(0, len(dados)):
        times.append(dados[indice].text)


    #capturando o placar da partida
    placar = []
    dados = driver.find_elements(By.CLASS_NAME, "sb-endstand")
    placar.append(dados[0].text[0])
    placar.append(dados[0].text[2])
    placar = [int(s) for s in placar]


    #catpura de dados estatísticos do jogo
    dados = driver.find_elements(By.CLASS_NAME, "sb-statistik-zahl")
    lista_dados = []
    for indice in range(0, len(dados)):
        lista_dados.append(dados[indice].text)
    lista_dados = [int(s) for s in lista_dados]
    
    #concatenação de todos os dados desse jogo
    dados_jogo = times + placar + lista_dados
    
    #adição dos dados desse jogo ao dataset
    dataset_jogos.loc[len(dataset_jogos)] = pd.Series(dados_jogo).values

    
    #funcionalidade para retornar à página inicial
    try:
        while(driver.title != titulo): #a variável titulo é o nome da página inicial
            driver.back()
            try:
                WebDriverWait(driver, 5).until( lambda driver: len(driver.title)>0 )
            except:    
                pass
    except:
        pass

    driver.get("https://www.transfermarkt.com.br/campeonato-brasileiro-serie-a/gesamtspielplan/wettbewerb/BRA1/saison_id/2024")
    indice = indice+1

IndexError: list index out of range

### **Testes**

Fluminense dentro de casa
média de gols: 1.35
finalizações: 16, 8, 14, 13, 21, 16, 9
escanteios: 5, 4, 4, 8, 6, 4, 3
faltas: 16, 10, 12, 12, 13, 24, 14

São Paulo fora de casa
media de gols: 0.94
finalizações: 14, 4, 8, 14, 10, 11, 6 
escanteios: 9, 3, 6, 3, 7, 5, 5
faltas: 9, 17, 9, 16, 17, 15, 17

In [33]:
fin_flu = [16, 8, 14, 13, 21, 16, 9]
esc_flu = [5, 4, 4, 8, 6, 4, 3]
fal_flu = [16, 10, 12, 12, 13, 24, 14]

fin_sp = [14, 4, 8, 14, 10, 11, 6 ]
esc_sp = [9, 3, 6, 3, 7, 5, 5]
fal_sp = [9, 17, 9, 16, 17, 15, 17]

In [ ]:
#analisando um duelo específico Cruzeiro x Corinthians
#veremos a media e moda dentro de casa do Cruzeiro, assim como a do Corinthians fora de casa

media_gols_cru_casa = resultados[resultados['time_casa'] == 'Fluminense']['gols_casa'].mean()
moda_gols_cru_casa = resultados[resultados['time_casa'] == 'Fluminense']['gols_casa'].mode()[0]

media_gols_cor_fora = resultados[resultados['time_visitante'] == 'São Paulo']['gols_visitante'].mean()
moda_gols_cor_fora = resultados[resultados['time_visitante'] == 'São Paulo']['gols_visitante'].mode()[0]

print("A MÉDIA de gols do Fluminense dentro de casa é {:.2f}".format(media_gols_cru_casa))
print("A MÉDIA de gols do São Paulo fora de casa é {:.2f}".format(media_gols_cor_fora))

In [39]:
print("Média de gols do Fluminense dentro de casa é {:.2f}".format(media_gols_cru_casa))
print("Média de finalizações do fluminense dentro de casa: {:.2f}".format( sum(fin_flu)/len(fin_flu) ))
print("Média de escanteios do fluminense dentro de casa: {:.2f}".format( sum(esc_flu)/len(esc_flu) ))
print("Média de faltas do fluminense dentro de casa: {:.2f}".format( sum(fal_flu)/len(fal_flu) ))

Média de gols do Fluminense dentro de casa é 1.35
Média de finalizações do fluminense dentro de casa: 13.86
Média de escanteios do fluminense dentro de casa: 4.86
Média de faltas do fluminense dentro de casa: 14.43


In [42]:
print("Média de gols do São Paulo fora de casa é {:.2f}".format(media_gols_cor_fora))
print("Média de finalizações do São Paulo fora de casa: {:.2f}".format( sum(fin_sp)/len(fin_sp) ))
print("Média de escanteios do São Paulo fora de casa: {:.2f}".format( sum(esc_sp)/len(esc_sp) ))
print("Média de faltas do São Paulo fora de casa: {:.2f}".format( sum(fal_sp)/len(fal_sp) ))

Média de gols do São Paulo fora de casa é 0.94
Média de finalizações do São Paulo fora de casa: 9.57
Média de escanteios do São Paulo fora de casa: 5.43
Média de faltas do São Paulo fora de casa: 14.29
